cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1823,10 +1451,79 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
                         kafkaCurrentReplicas = 0;
                         if (sts != null && sts.getSpec() != null)   {
                             kafkaCurrentReplicas = sts.getSpec().getReplicas();
+                            this.currentVersion = Annotations.annotations(sts).get(ANNO_STRIMZI_IO_KAFKA_VERSION);
                         }
 
                         this.kafkaCluster = KafkaCluster.fromCrd(kafkaAssembly, versions, oldStorage, kafkaCurrentReplicas);
                         this.kafkaBootstrapDnsName.addAll(ListenersUtils.alternativeNames(kafkaCluster.getListeners()));
+
+                        //return Future.succeededFuture(this);
+                        return podOperations.listAsync(namespace, this.kafkaCluster.getSelectorLabels());
+                    }).compose(pods -> {
+                        String lowestKafkaVersion = currentVersion;
+                        String highestKafkaVersion = currentVersion;
+
+                        for (Pod pod : pods)    {
+                            // Collect the different annotations from the pods
+                            String kafkaVersionAnnotation = Annotations.stringAnnotation(pod, ANNO_STRIMZI_IO_KAFKA_VERSION, null);
```

Comment:
I think names like `currentVersion`, `currentMessageFormat` and `currentIbp` would make this bit of code clearer.